test/abe/dippe.c

```diff
@@ -0,0 +1,214 @@
+
+#include <amcl/pair_BN254.h>
+
+#include "cifer/test.h"
+#include "cifer/data/vec.h"
+#include "cifer/abe/dippe.h"
+
+MunitResult test_dippe_end_to_end_conjunction(const MunitParameter *params, void *data) {
+    cfe_error err;
+
+    // Init DIPPE scheme
+    cfe_dippe dippe;
+    cfe_dippe_setup(&dippe, 2);
+
+    // Setup two Authorities
+    cfe_dippe_pub_key pk[2];
+    cfe_dippe_sec_key sk[2];
+    cfe_dippe_authsetup(&dippe, &(pk[0]), &(sk[0]));
+    cfe_dippe_authsetup(&dippe, &(pk[1]), &(sk[1]));
+
+    // Sample message
+    ECP_BN254 g1;
+    ECP_BN254_generator(&g1);
+    ECP2_BN254 g2;
+    ECP2_BN254_generator(&g2);
+    FP12_BN254 msg;
+    PAIR_BN254_ate(&msg, &g2, &g1);
+    PAIR_BN254_fexp(&msg);
+
+    // Collection of 6 public keys
+    // Each of them correspondents to an index of the Attribute-/ Policy-vector
+    cfe_dippe_pub_key *pks[] = {&(pk[0]), &(pk[1]), &(pk[0]), &(pk[1]), &(pk[0]), &(pk[1])};
+
+    // unique GID
+    char gid[] = "TESTGID";
+
+    // Test patterns
+    char attr_patterns[4][6] = {
+        "11011", "11111",   // valid patterns
+        "01001", "11010"    // invalid patterns
+    };
+    // Test results
+    int decryption_results[] = {1, 1, 0, 0};
+
+    // Policy vector (one additional component is added)
+    //   ------            -----
+    //   Policy            Attrs
+    //   ------            -----
+    //    x0               1
+    //    x1               1
+    //    0         .      0      =  (x0+x1+x4-x0-x1-x4) = 0
+    //    0                0
+    //    x4               1
+    // -> x5=(-x0-x1-x4)   1
+    cfe_vec pol;
+    err = cfe_dippe_build_conjunction_policy_vector(&dippe, &pol, "11001");
+    munit_assert(err == CFE_ERR_NONE);
+
+    mpz_t     attrs;
+    cfe_dippe_cipher cipher;
+
+    // Decrypt message
+    FP12_BN254 result;
+
+    // Collection of user secret keys
+    cfe_dippe_user_secret_key usks[6];
+
+    for (int i=0; i<4; i++) {
+        // Attribute vector (one additional component is added)
+        err = cfe_dippe_build_conjunction_attribute_vector(&dippe, &attrs, attr_patterns[i]);
+        munit_assert(err == CFE_ERR_NONE);
+
+        // Encrypt message under given policy vector
+        err = cfe_dippe_encrypt(&dippe, &cipher, pks, (sizeof(pks)/sizeof(cfe_dippe_pub_key*)), &pol, &msg);
```

Comment:
I think you should put encryption before the loop. Encryption is done once, then different users try to decrypt it with different attributes.

Reply:
You're right, I moved the encryption algorithm outside the loop.